File: src/scripting/api/assets/texture.cpp
// Copyright (c) 2024 Mario "Neo" Sieg. All Rights Reserved.

#include "../_prelude.hpp"

using graphics::texture;

LUA_INTEROP_API auto __lu_texture_load(const char* const resource_path) -> assetmgr::interop_asset_id {
    return scene_mgr::active().get_asset_registry<texture>().interop.load(resource_path);
}

LUA_INTEROP_API auto __lu_texture_get_width(const assetmgr::interop_asset_id id) -> std::uint32_t {
    get_resource_property(texture, width(), 0);
}

LUA_INTEROP_API auto __lu_texture_get_height(const assetmgr::interop_asset_id id) -> std::uint32_t {
    get_resource_property(texture, height(), 0);
}

LUA_INTEROP_API auto __lu_texture_get_depth(const assetmgr::interop_asset_id id) -> std::uint32_t {
    get_resource_property(texture, depth(), 0);
}

LUA_INTEROP_API auto __lu_texture_get_mip_levels(const assetmgr::interop_asset_id id) -> std::uint32_t {
    get_resource_property(texture, mip_levels(), 0);
}

LUA_INTEROP_API auto __lu_texture_get_array_size(const assetmgr::interop_asset_id id) -> std::uint32_t {
    get_resource_property(texture, array_size(), 0);
}

LUA_INTEROP_API auto __lu_texture_is_cubemap(const assetmgr::interop_asset_id id) -> bool {
    get_resource_property(texture, is_cubemap(), false);
}

LUA_INTEROP_API auto __lu_texture_get_format(const assetmgr::interop_asset_id id) -> std::uint32_t {
    get_resource_property(texture, format_int(), 0);
}

File: engine_assets/scripts/core/assets/texture.lua
-- Copyright (c) 2024 Mario "Neo" Sieg. All Rights Reserved.

----------------------------------------------------------------------------
--- texture Module - Implements the texture asset.
--- @module texture
------------------------------------------------------------------------------

local ffi = require 'ffi'
local cpp = ffi.C
local formats = require 'detail.texture_format'

ffi.cdef [[
    __asset_id __lu_texture_load(const char* path);
    uint32_t __lu_texture_get_width(__asset_id id);
    uint32_t __lu_texture_get_height(__asset_id id);
    uint32_t __lu_texture_get_depth(__asset_id id);
    uint32_t __lu_texture_get_mip_levels(__asset_id id);
    uint32_t __lu_texture_get_array_size(__asset_id id);
    bool __lu_texture_is_cubemap(__asset_id id);
    uint32_t __lu_texture_get_format(__asset_id id);
]]

local invalid_id = 0x7fffffff

local texture = {
    _id = 0, --  asset id (scene local)
    _path = '' -- asset path
}

function texture:get_id() return self._id end
function texture:get_path() return self._path end
function texture:get_width() return cpp.__lu_texture_get_width(self._id) end
function texture:get_height() return cpp.__lu_texture_get_height(self._id) end
function texture:get_depth() return cpp.__lu_texture_get_depth(self._id) end
function texture:get_mip_levels() return cpp.__lu_texture_get_mip_levels(self._id) end
function texture:get_array_size() return cpp.__lu_texture_get_array_size(self._id) end
function texture:is_cubemap() return cpp.__lu_texture_is_cubemap(self._id) end
function texture:get_format() return formats.formats_by_id[cpp.__lu_texture_get_format(self._id)] end
function texture:__tostring() return string.format('texture: %s', self._path) end

function texture:from_id(id, path)
    assert(type(id) == 'number' and id ~= invalid_id)
    local o = {}
    setmetatable(o, { __index = self })
    o._id = id
    o._path = path
    return o
end

function texture:load(path)
    assert(type(path) == 'string' and path ~= '')
    local id = cpp.__lu_texture_load(path)
    if id == invalid_id then return nil end
    return self:from_id(id, path)
end

return texture
